src/software/jetson_nano/thunderloop.cpp

```diff
@@ -96,51 +129,69 @@ void Thunderloop::runLoop()
 
                 // Update primitive executor's primitive set
                 {
-                    ScopedTimespecTimer timer(&poll_time);
-                    primitive_executor_.updatePrimitiveSet(robot_id_, primitive_set_);
+                    clock_gettime(CLOCK_MONOTONIC, &last_primitive_received_time);
+
+                    // Start new primitive
+                    {
+                        ScopedTimespecTimer timer(&poll_time);
+                        primitive_executor_.updatePrimitiveSet(robot_id_, primitive_set_);
+                    }
+
+                    thunderloop_status_.set_primitive_executor_start_time_ns(
+                        static_cast<unsigned long>(poll_time.tv_nsec));
                 }
+            }
 
-                thunderloop_status_.set_primitive_executor_start_time_ns(
-                    static_cast<unsigned long>(poll_time.tv_nsec));
+            // If the world msg is new, update the internal buffer
+            if (new_world.time_sent().epoch_timestamp_seconds() >
+                world_.time_sent().epoch_timestamp_seconds())
+            {
+                primitive_executor_.updateWorld(new_world);
```

Comment:
It doesn't seem like we ever call `primitive_executor_.updateLocalVelocity` with the current local velocity measured vy the encoders. Would that be a part of a future PR, or can it be added here aswell?

Reply:
yup! added here

i was under the impression that its unused, but I added it because its quick